core/dbt/include/global_project/macros/adapters/columns.sql

```diff
@@ -17,23 +17,60 @@
 {% endmacro %}
 
 
+{% macro get_empty_subquery_sql(select_sql) -%}
+  {{ return(adapter.dispatch('get_empty_subquery_sql', 'dbt')(select_sql)) }}
+{% endmacro %}
+
+{% macro default__get_empty_subquery_sql(select_sql) %}
+    select * from (
+        {{ select_sql }}
+    ) as __dbt_sbq
+    where false
+    limit 0
+{% endmacro %}
+
+
+{% macro get_empty_schema_sql(columns) -%}
+  {{ return(adapter.dispatch('get_empty_schema_sql', 'dbt')(columns)) }}
+{% endmacro %}
+
+{% macro default__get_empty_schema_sql(columns) %}
+    select
+    {% for i in columns %}
+      {%- set col = columns[i] -%}
+      cast(null as {{ col['data_type'] }}) as {{ col['name'] }}{{ ", " if not loop.last }}
+    {%- endfor -%}
+{% endmacro %}
+
+{% macro get_column_schema_from_query(select_sql) -%}
+  {{ return(adapter.dispatch('get_column_schema_from_query', 'dbt')(select_sql)) }}
+{% endmacro %}
+
+{% macro default__get_column_schema_from_query(select_sql) %}
```

Comment:
I don't like this name. I'm not sure why. I feel like its saying "get column schema from some query" when it's far more deliberate than that in practice, as I understand it. 

This macro is said to "Get a list of the column names and data types from the given sql." What is that given SQL? Why is it? These are questions my brain latches onto as though part of the abstraction that'd help explain this is missing.

Reply:
> What is that given SQL? Why is it? These are questions my brain latches onto as though part of the abstraction that'd help explain this is missing.

Agree - but I do want to avoid an overly-specific name to this more general function. The deliberate use case here is to both 'get a list of column names and data types based on the schema yml' and 'get a list of column names and data types based on the user-provided SQL'. Both use cases call this underlying method `get_column_schema_from_query`, and presumably could be used for other applications in the future! 

I went with `get_column_schema..` here as opposed to `get_columns..` because the method is not returning the `Column` class but a list of tuples representing the schema. Don't have a strong preference between `from_query` or `from_sql`. I think I just went with `from_query` for consistency with the existing `get_columns_from_query` adapter method. 

I'll add some more comments here in terms of the deliberate use case in [assert_columns_equivalent](https://github.com/dbt-labs/dbt-core/pull/6986/files#diff-28fef0f8080904ca9d7c04e61050beded83aaecc39d332b8f0486eea686d44f1R31-R32), where we're composing these more generic methods to get a very deliberate and narrow functionality.